Sources/NIO/Bootstrap.swift

```diff
@@ -364,8 +385,121 @@ public final class ServerBootstrap {
+        /// - Returns: the modified bootstrap (currently the same one mutated)
+        func applyOption(to serverBootstrap: ServerBootstrap) {
+            data.applyOption(to: serverBootstrap)
+        }
+        
+        fileprivate enum ShorthandServerOption {
+            case reuseAddr(Bool)
+            case autoRead(Bool)
+            case backlog(Int32)
+            
+            func applyOption(to serverBootstrap: ServerBootstrap) {
+                switch self {
+                case .autoRead(let value):
+                    _ = serverBootstrap.serverChannelOption(ChannelOptions.autoRead, value: value)
+                case .reuseAddr(let value):
+                    _ = serverBootstrap.serverChannelOption(ChannelOptions.socketOption(.reuseaddr),
+                                                            value: value ? 1 : 0)
+                case .backlog(let value):
+                    _ = serverBootstrap.serverChannelOption(ChannelOptions.backlog, value: value)
+                }
+            }
+        }
+    }
+    
+    /// A channel option which can be applied to bootstrap using shorthand notation.
+    /// - See: ServerBootstrap.serverOptions(_ options: [ShorthandBootstrapOption])
+    public struct ShorthandChildBootstrapOption {
+        private let data: ShorthandChildOption
+        
+        private init(_ data: ShorthandChildOption) {
+            self.data = data
+        }
+        
+        /// Apply the contained option to the supplied ServerBootstrap
+        /// - Parameter serverBootstrap: bootstrap to apply this option to.
+        /// - Returns: the modified bootstrap (currently the same one mutated)
+        func applyOption(to serverBootstrap: ServerBootstrap) {
+            data.applyOption(to: serverBootstrap)
+        }
+        
+        fileprivate enum ShorthandChildOption {
+            case reuseAddr(Bool)
+            case allowRemoteHalfClosure(Bool)
+            
+            func applyOption(to serverBootstrap: ServerBootstrap) {
+                switch self {
+                case .reuseAddr(let value):
+                    _ = serverBootstrap.childChannelOption(ChannelOptions.socketOption(.reuseaddr),
+                                                           value: value ? 1 : 0)
+                case .allowRemoteHalfClosure(let value):
+                    _ = serverBootstrap.childChannelOption(ChannelOptions.allowRemoteHalfClosure, value: value)
+                }
+            }
+        }
+    }
+}
+
+// Hashable for the convenience of users.
+extension ServerBootstrap.ShorthandServerBootstrapOption : Hashable {}
+extension ServerBootstrap.ShorthandServerBootstrapOption.ShorthandServerOption : Hashable {}
+
+/// Approved shorthand server options.
+extension ServerBootstrap.ShorthandServerBootstrapOption {
+    /// Option to reuse address.
+    /// - See:  NIOBSDSocket.Option.reuseaddr
+    public static let reuseAddr = ServerBootstrap.ShorthandServerBootstrapOption(.reuseAddr(true))
```

Comment:
Given this is the higher level API for configuring options: should we name this `reuseAddresses` instead?

Reply:
I would even add more words. Something like `.allowImmediateEnpointAddressReuse` or so.

Reply:
@Lukasa any good naming suggestions?

Reply:
I like yours.

Reply:
@Lukasa also: wasn't there some impedance mismatch between Network.framework and `SO_REUSEADDR`?

Reply:
Yes, network.framework only has one that blurs the line between `SO_REUSEADDR` and `SO_REUSEPORT`: [`allowLocalEndpointReuse`](https://developer.apple.com/documentation/network/nwparameters/2998692-allowlocalendpointreuse).

Reply:
@Lukasa hmm, are you sure?

This program

```swift
import Network

var parameters = NWParameters.tcp
parameters.allowLocalEndpointReuse = true
let port: NWEndpoint.Port = NWEndpoint.Port(rawValue: 12345)!
let listener = try NWListener(using: parameters, on: port)

listener.stateUpdateHandler = { state in
    print("state", state)
}
listener.newConnectionHandler = { conn in
    print("conn", conn)
}
listener.start(queue: DispatchQueue.main)
dispatchMain()
listener.cancel()
```

when started twice gives me

```
state failed(POSIXErrorCode: Address already in use)
```

on the second attempt. So it seems that `allowLocalEndpointReuse` == `SO_REUSEADDR`?

Reply:
and not `SO_REUSEPORT`

Reply:
The answer is that I am not sure. The description leaves it fundamentally vague.

Reply:
I reckon they just don't support `SO_REUSEPORT` at all. Doesn't properly work on Darwin anyway. So I think we can just treat it like `SO_REUSEADDR`.

Reply:
reuseAddr -> allowImmediateEndpointAddressReuse